### PR TITLE
feat: [CDS-44943]: execution panel grid/list view + search panel

### PR DESCRIPTION
**Pull request by PratyushGar**
#### Screenshots

https://user-images.githubusercontent.com/100121280/198268103-39e8b4f2-ec52-441c-a2bf-ae77283507cb.mov

image
image

## **Search steps**

image
image

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelUtils.tsx [hunk 1]
```diff
@@ -14,14 +14,19 @@ import type { TemplateDetailsByRef } from '@cd/context/DeploymentContext/Deploym
...
   deploymentConfig: DeploymentConfig
+  showWarning: (message: React.ReactNode, timeout?: number | undefined, key?: string | undefined) => void
 }
 
-export const getUpdatedDeploymentConfig = ({ templateRef, deploymentConfig }: Params): DeploymentConfig =>
+export const getUpdatedDeploymentConfig = ({ templateRef, deploymentConfig, showWarning }: Params): DeploymentConfig =>
   produce(deploymentConfig, draft => {
     const stepTemplateRefs = deploymentConfig.execution?.stepTemplateRefs || /* istanbul ignore next */ []
-    const updatedStepTemplateRefs = [templateRef, ...stepTemplateRefs]
+    if (stepTemplateRefs.some(item => item === templateRef)) {
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelUtils.tsx:23]: This function is not supposed to show warning from it's name. Refractor this.

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelUtils.tsx [hunk 2]
```diff
@@ -14,14 +14,19 @@ import type { TemplateDetailsByRef } from '@cd/context/DeploymentContext/Deploym
...
+      showWarning('Duplicate step cannot be added')
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelUtils.tsx:24]: This function is not supposed to show warning from it's name. Refractor this.
Also move string to string yaml

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelListView.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2022 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Shield 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+import React from 'react'
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanelListView.tsx:8]: We should have used the same list view. Any change in the template list will require double effort

**Review by samarth-sharma31 — DISMISSED** [review]

File: src/modules/72-templates-library/pages/TemplatesPage/views/TemplatesView/TemplatesView.tsx [hunk 4]
```diff
@@ -22,6 +22,7 @@ export interface TemplatesViewProps {
   onOpenEdit?: (template: TemplateSummaryResponse) => void
   onOpenSettings?: (templateIdentifier: string) => void
   onDelete?: (template: TemplateSummaryResponse) => void
+  enablePagination?: boolean
```
Comment (yc147) [inline src/modules/72-templates-library/pages/TemplatesPage/views/TemplatesView/TemplatesView.tsx:25]: This is not required anymore.

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx [hunk 5]
```diff
@@ -43,12 +62,30 @@ export function ExecutionPanel({ children }: React.PropsWithChildren<unknown>):
...
 
-  const { getString } = useStrings()
+  const reset = (): void => {
+    searchRef.current?.clear()
+    setSearchString(undefined)
+    setSelectedTemplate(undefined)
+  }
+
+  useEffect(() => {
+    setStepTemplateRefsList(stepTemplateRefs)
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx:86]: Move this line to rest function

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx [hunk 6]
```diff
@@ -107,39 +141,169 @@ export function ExecutionPanel({ children }: React.PropsWithChildren<unknown>):
...
+  }
+
+  const handleCardRemove = (template: TemplateSummaryResponse): void => {
+    const updatedDeploymentConfig = produce(deploymentConfig, draft => {
+      const updatedStepTemplateRefs = stepTemplateRefs.filter(
+        templateRef => templateRef !== getScopeBasedTemplateRef(template)
+      )
+      set(draft, 'execution.stepTemplateRefs', updatedStepTemplateRefs)
+    })
+    showSuccess(getString('cd.removeStepTemplateSuccess'))
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx:180]: Ideally you show call this on successfully resolution of `updateDeploymentConfig` promise

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx [hunk 7]
```diff
@@ -107,39 +141,169 @@ export function ExecutionPanel({ children }: React.PropsWithChildren<unknown>):
...
+        enablePagination={false}
+      />
+    )
+  }, [stepTemplateRefsList, view, handleCardClick, templateDetailsByRef, selectedTemplate])
+
+  const onChange = (text: string): void => {
+    let filteredstepTemplateRefs = [...stepTemplateRefs]
+    const searchVal = text.trim()
+    setSearchString(searchVal)
+    if (searchVal) {
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx:250]: use `useEffect` on `searchString` to set `stepTemplateRefsList`

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/DTListCardContextMenu/DTListCardContextMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  onDelete: (template: TemplateSummaryResponse) => void
+  className?: string
+}
+
+export const DTListCardContextMenu: React.FC<ContextMenuProps> = (props): JSX.Element => {
+  const { getString } = useStrings()
+  const { template, onPreview, onOpenEdit, onDelete, className, ...popoverProps } = props
+  const [menuOpen, setMenuOpen] = React.useState(false)
+  const { accountId, orgIdentifier, projectIdentifier, templateIdentifier } = useParams<TemplateStudioPathProps>()
+  const [canView, canEdit, canDelete] = usePermission(
```
Comment (yc147) [inline src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/DTListCardContextMenu/DTListCardContextMenu.tsx:38]: Permission logic is wrong

File: src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx (hunk 5 above)
Comment (yc147) [reply src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx:86]: Ideally you just need to reset search box when resetting.

Comment (PratyushGar) [reply src/modules/75-cd/components/TemplateStudio/DeploymentTemplateCanvas/DeploymentTemplateForm/components/ExecutionPanel/ExecutionPanel.tsx]: Done 👍

**Review by samarth-sharma31 — APPROVED** [review]

**Review by YounixIO — APPROVED** [review]

**Pull request merged**